sdk/keyvault/azure-keyvault-keys/azure/keyvault/keys/crypto/_models.py

```diff
@@ -2,11 +2,432 @@
+        :param padding: The padding to use. Supported paddings are `PKCS1v15` and `PSS`. For `PSS`, the only supported
+            mask generation function is `MGF1`. See https://learn.microsoft.com/azure/key-vault/keys/about-keys-details
+            for details.
+        :type padding: :class:`~cryptography.hazmat.primitives.asymmetric.padding.AsymmetricPadding`
+        :param algorithm: The algorithm to sign with. Only `HashAlgorithm`s are supported -- specifically, `SHA256`,
+            `SHA384`, and `SHA512`.
+        :type algorithm: :class:`~cryptography.hazmat.primitives.asymmetric.utils.Prehashed` or
+            :class:`~cryptography.hazmat.primitives.hashes.HashAlgorithm`
+
+        :returns: The signature, as bytes.
+        :rtype: bytes
+        """
+        if isinstance(algorithm, Prehashed):
+            raise ValueError("`Prehashed` algorithms are unsupported. Please provide a `HashAlgorithm` instead.")
+        mapped_algorithm = get_signature_algorithm(padding, algorithm)
+        digest = Hash(algorithm)
+        digest.update(data)
+        result = self._client.sign(mapped_algorithm, digest.finalize())
+        return result.signature
+
+    def private_numbers(self) -> RSAPrivateNumbers:
+        """Returns an `RSAPrivateNumbers` representing the key's private numbers.
+
+        :returns: The private numbers of the key.
+        :rtype: :class:`~cryptography.hazmat.primitives.asymmetric.rsa.RSAPrivateNumbers`
+        """
+        # Fetch public numbers from JWK
+        e = int.from_bytes(self._key.e, "big")  # type: ignore[attr-defined]
+        n = int.from_bytes(self._key.n, "big")  # type: ignore[attr-defined]
+        public_numbers = RSAPublicNumbers(e, n)
+
+        # Fetch private numbers from JWK
+        p = int.from_bytes(self._key.p, "big") if self._key.p else None  # type: ignore[attr-defined]
+        q = int.from_bytes(self._key.q, "big") if self._key.q else None  # type: ignore[attr-defined]
+        d = int.from_bytes(self._key.d, "big") if self._key.d else None  # type: ignore[attr-defined]
+        dmp1 = int.from_bytes(self._key.dp, "big") if self._key.dp else None  # type: ignore[attr-defined]
+        dmq1 = int.from_bytes(self._key.dq, "big") if self._key.dq else None  # type: ignore[attr-defined]
+        iqmp = int.from_bytes(self._key.qi, "big") if self._key.qi else None  # type: ignore[attr-defined]
+
+        # Calculate any missing attributes
+        if d is None:
+            raise ValueError("An 'RSAPrivateNumbers' couldn't be created with the available key material.")
+        if p is None or q is None:
+            p, q = rsa_recover_prime_factors(n, e, d)
+        if dmp1 is None:
+            dmp1 = rsa_crt_dmp1(d, p)
+        if dmq1 is None:
+            dmq1 = rsa_crt_dmq1(d, q)
+        if iqmp is None:
+            iqmp = rsa_crt_iqmp(p, q)
+
+        return RSAPrivateNumbers(p, q, d, dmp1, dmq1, iqmp, public_numbers)
+
+    def private_bytes(
```

Comment:
same comment as above.

Reply:
Good point; we should probably raise a more specific error in this method if we don't have the necessary information. This method tries to get private numbers first, at which point we'd get
```
ValueError: An 'RSAPrivateNumbers' couldn't be created with the available key material.
```
This would be confusing to raise from `private_bytes` without implementation context.